File: python/caesar_cipher.py
import re

def caesar_cipher(string, index):
    alpha_low = 'abcdefghijklmnopqrstuvwxyz'
    alpha_up = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
    result = []
    i = 0
    upper_case = False
    while i < len(string):
        str_char = str(string[i])
        if re.match("[a-z]", str_char):
            letter_index = alpha_low.find(str_char)
            if (letter_index + index) < 0:
                letter_index += 26
                result.append(alpha_low[letter_index + index])
            elif (letter_index + index) > 26:
                letter_index -= 26
                result.append(alpha_low[letter_index + index])
            else:
                result.append(alpha_low[letter_index + index])
        elif re.match("[A-Z]", str_char):
            letter_index = alpha_up.find(str_char) 
            if (letter_index + index) < 0:
                letter_index += 26
                result.append(alpha_up[letter_index + index])
            elif (letter_index + index) > 26:
                letter_index -= 26
                result.append(alpha_up[letter_index + index])
            else:
                result.append(alpha_up[letter_index + index])
        elif re.match("[^a-zA-Z]", str_char):
            result.append(str_char)           
        i += 1

    return ''.join(result)

print(caesar_cipher("What a string!", 5))

        # if re.match("[^a-zA-Z]", string[i]):
        #     result.append(string[i])
        #     i += 1
        # else:
        #     re.match("[a-z]", string[i])
        #     lower_case = True
        #     letter_index = alpha_low.index(string[i])
        #     print(letter_index)
        #     i += 1

File: js/caesarCipher.js
exports.caesarCipher = function(string, index) {
    let alphaLow = 'abcdefghijklmnopqrstuvwxyz';
    let alphaUp = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ';
    let result = [];
    let i = 0;
    // iterate over string
    while (i < string.length){
      // if not a letter
      // could also do regex
      if (((alphaLow.indexOf(string[i]) === -1)) && ((alphaUp.indexOf(string[i]) === -1))){
        result.push(string[i]);
        i++;
        continue;
      }
      // Could also set an uppercase = True/False and use only one block of code
      // easier to do with charCode
      // if lowercase
      if (string[i] !== string[i].toUpperCase()){
          let letterIndex = alphaLow.indexOf(string[i]);
          // if shifted letter index is less than zero, add 26
        if ((letterIndex + index) < 0){
          letterIndex += 26;
          result.push(alphaLow[letterIndex + index]);
          // if shifted letter index is greater than 26, subtract 26
        } else if ((letterIndex + index) > 26){
          letterIndex -= 26
          result.push(alphaLow[letterIndex + index]);
          // if shifted index is between 0 and 26, use original index
        } else {
          result.push(alphaLow[letterIndex + index]);
        } 
      }
      // if uppercase	
      if (string[i] === string[i].toUpperCase()){
          let letterIndex = alphaUp.indexOf(string[i]);
        if ((letterIndex + index) < 0){
          letterIndex += 26;
          result.push(alphaUp[letterIndex + index]);
        } else if ((letterIndex + index) > 26){
          letterIndex -= 26
          result.push(alphaUp[letterIndex + index]);
        } else {
          result.push(alphaUp[letterIndex + index]);
        } 
      }
      i++;
    }
    return result.join('');
};
//console.log(exports.caesarCipher("What a string!", 5))